From 9179cb668b3186b2ffd336dd9f04ca53d813b852 Mon Sep 17 00:00:00 2001
From: Hari Naraayana Desikan Kannan <hnaraaya@qti.qualcomm.com>
Date: Thu, 10 Jul 2025 09:55:48 +0530
Subject: [PATCH 3/3] UPSTREAM:PROTOCOL: wifi: ath12k: add WMI support for
 station link reconfiguration

Add support for handling associated station link reconfiguration by
updating firmware through WMI peer association commands.

Extend the peer association command to encode the following fields:
- ml_reconfig: indicates a reconfiguration event
- mlo_link_add / mlo_link_del: specify per-link additions or removals

Update ath12k_peer_assoc_h_mlo() to set the ml_reconfig, mlo_link_add,
and mlo_link_del flags in the peer_assoc_mlo_params structure based on
the reconf.added_links and reconf.removed_links bitmaps provided by
mac80211.

These flags are applied to both the self and partner links involved in
the reconfiguration. This allows firmware to interpret the context and
distinguish dynamic link updates from initial association.

Firmware uses these flags to block Tx/Rx on newly added links until
peer authorization is complete, ensure correct partner link setup
during transition, and support simultaneous link additions and
deletions while preserving MLD peer integrity.

Signed-off-by: Hari Naraayana Desikan Kannan <hnaraaya@qti.qualcomm.com>
---
 drivers/net/wireless/ath/ath12k/mac.c | 10 ++++++++++
 drivers/net/wireless/ath/ath12k/wmi.c | 12 ++++++++++++
 drivers/net/wireless/ath/ath12k/wmi.h | 23 +++++++++++++++++++++++
 3 files changed, 45 insertions(+)

--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -4205,6 +4205,12 @@ static void ath12k_peer_assoc_h_mlo(stru
 	ml->num_partner_links = 0;
 	links = ahsta->links_map;
 
+	if (sta->reconf.removed_links & BIT(arsta->link_id))
+		ml->ml_reconfig = ml->mlo_link_del = true;
+
+	if (sta->reconf.added_links & BIT(arsta->link_id))
+		ml->ml_reconfig = ml->mlo_link_add = true;
+
 	rcu_read_lock();
 
 	i = 0;
@@ -4232,6 +4238,10 @@ static void ath12k_peer_assoc_h_mlo(stru
 			   ml->partner_info[i].primary_umac = false;
 		   ml->partner_info[i].logical_link_idx_valid = true;
 		ml->partner_info[i].logical_link_idx = arsta_p->link_idx;
+		if (sta->reconf.removed_links & BIT(arsta_p->link_id))
+			ml->ml_reconfig = ml->partner_info[i].mlo_link_del = true;
+		if (sta->reconf.added_links & BIT(arsta_p->link_id))
+			ml->ml_reconfig = ml->partner_info[i].mlo_link_add = true;
 		ml->num_partner_links++;
 
 		i++;
@@ -20906,6 +20916,7 @@ static int ath12k_mac_hw_register(struct
 	int ret, i, j;
 	u32 ht_cap = U32_MAX, antennas_rx = 0, antennas_tx = 0;
 	bool is_6ghz = false, is_raw_mode = false, is_monitor_disable = false;
+	bool is_sta_ml_recfg = false;
 	u8 *mac_addr = NULL;
 	u8 mbssid_max_interfaces = 0;
 
@@ -20947,6 +20958,9 @@ static int ath12k_mac_hw_register(struct
 		if (test_bit(ATH12K_GROUP_FLAG_RAW_MODE, &ar->ab->ag->flags))
 			is_raw_mode = true;
 
+		if (test_bit(WMI_SERVICE_STA_MLO_RCFG_SUPPORT, ar->ab->wmi_ab.svc_map))
+			is_sta_ml_recfg = true;
+
 		if (!ar->ab->hw_params->supports_monitor)
 			is_monitor_disable = true;
 
@@ -21071,6 +21085,10 @@ static int ath12k_mac_hw_register(struct
 
 		if(!is_raw_mode)
 			ieee80211_hw_set(hw, MLO_MCAST_MULTI_LINK_TX);
+
+		if(is_sta_ml_recfg)
+			ath12k_iftypes_ext_capa[2].mld_capa_and_ops |=
+				IEEE80211_MLD_CAP_OP_LINK_RECONF_SUPPORT;
 	}
 
 	hw->queues = ATH12K_HW_MAX_QUEUES;
--- a/drivers/net/wireless/ath/ath12k/wmi.c
+++ b/drivers/net/wireless/ath/ath12k/wmi.c
@@ -3182,11 +3182,21 @@ int ath12k_wmi_send_peer_assoc_cmd(struc
 	if (arg->ml.bridge_peer)
 		ml_params->flags |= cpu_to_le32(ATH12K_WMI_FLAG_MLO_BRIDGE_PEER);
 
+	if (arg->ml.mlo_link_add)
+		ml_params->flags |= cpu_to_le32(ATH12K_WMI_FLAG_MLO_LINK_ADD);
+
+	if (arg->ml.mlo_link_del)
+		ml_params->flags |= cpu_to_le32(ATH12K_WMI_FLAG_MLO_LINK_DEL);
+
 	ether_addr_copy(ml_params->mld_addr.addr, arg->ml.mld_addr);
 	ml_params->logical_link_idx = cpu_to_le32(arg->ml.logical_link_idx);
 	ml_params->ml_peer_id = cpu_to_le32(arg->ml.ml_peer_id);
 	ml_params->ieee_link_id = cpu_to_le32(arg->ml.ieee_link_id);
 
+	if (arg->ml.ml_reconfig)
+		ml_params->ml_reconfig = 1;
+	else
+		ml_params->ml_reconfig = 0;
 	eml_cap = arg->ml.eml_cap;
 	if (u16_get_bits(eml_cap, IEEE80211_EML_CAP_EMLSR_SUPP)) {
 		/* Padding delay */
@@ -3270,6 +3280,12 @@ skip_ml_params:
 			partner_info->flags |=
 				cpu_to_le32(ATH12K_WMI_FLAG_MLO_BRIDGE_PEER);
 
+		if (arg->ml.partner_info[i].mlo_link_add)
+			partner_info->flags |= cpu_to_le32(ATH12K_WMI_FLAG_MLO_LINK_ADD);
+
+		if (arg->ml.partner_info[i].mlo_link_del)
+			partner_info->flags |= cpu_to_le32(ATH12K_WMI_FLAG_MLO_LINK_DEL);
+
 		partner_info->logical_link_idx =
 			cpu_to_le32(arg->ml.partner_info[i].logical_link_idx);
 		ptr += sizeof(*partner_info);
--- a/drivers/net/wireless/ath/ath12k/wmi.h
+++ b/drivers/net/wireless/ath/ath12k/wmi.h
@@ -2514,6 +2514,7 @@ enum wmi_tlv_service {
 	WMI_SERVICE_THERM_THROT_5_LEVELS = 429,
 
 	WMI_SERVICE_UMAC_MIGRATION_SUPPORT = 436,
+	WMI_SERVICE_STA_MLO_RCFG_SUPPORT = 448,
 
 	WMI_MAX_EXT2_SERVICE,
 };
@@ -3271,6 +3272,7 @@ struct wmi_vdev_create_mlo_params {
 #define ATH12K_WMI_FLAG_MLO_EMLSR_SUPPORT		BIT(6)
 #define ATH12K_WMI_FLAG_MLO_FORCED_INACTIVE		BIT(7)
 #define ATH12K_WMI_FLAG_MLO_LINK_ADD			BIT(8)
+#define ATH12K_WMI_FLAG_MLO_LINK_DEL			BIT(9)
 #define ATH12K_WMI_FLAG_MLO_BRIDGE_PEER			BIT(10)
 #define ATH12K_WMI_FLAG_MLO_BRIDGE_LINK			BIT(14)
 #define ATH12K_WMI_FLAG_MLO_IEEE_LINK_IDX_VALID		BIT(18)
@@ -3470,6 +3472,8 @@ struct wmi_ml_partner_info {
 	bool mlo_bridge_link;
 	bool bridge_peer;
 	u32 logical_link_idx;
+	bool mlo_link_add;
+	bool mlo_link_del;
 };
 
 struct wmi_ml_arg {
@@ -4355,6 +4359,9 @@ struct peer_assoc_mlo_params {
 	u8 num_partner_links;
 	struct wmi_ml_partner_info partner_info[ATH12K_WMI_MLO_PEER_MAX_LINKS];
 	u16 eml_cap;
+	bool ml_reconfig;
+	bool mlo_link_add;
+	bool mlo_link_del;
 };
 
 struct wmi_rate_set_arg {
@@ -4610,6 +4617,23 @@ struct wmi_peer_assoc_mlo_params {
 	__le32 emlsr_trans_timeout_us;
 	__le32 emlsr_trans_delay_us;
 	__le32 emlsr_padding_delay_us;
+	union {
+		u32 msd_dur_subfield;
+		u32 msd_dur_us;
+	};
+	u32 msd_ofdm_ed_thr;
+	u32 msd_max_num_txops;
+	u32 max_num_simultaneous_links;
+	u32 nstr_indication_bitmap;
+	u32 recommended_max_num_simultaneous_links;
+
+	union {
+		u32 ml_reconfig__word;
+		struct {
+			u32 ml_reconfig: 1,
+			    unused: 31;
+		};
+	};
 } __packed;
 
 struct wmi_peer_assoc_complete_cmd {
